main/src/ca/uwaterloo/flix/language/ast/Purity.scala

```diff
@@ -0,0 +1,36 @@
+/*
+ * Copyright 2022 Christian Bonde, Patrick Lundvig, Anna Krogh
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package ca.uwaterloo.flix.language.ast
+
+sealed trait Purity
+
+
+/**
+ * Purity of an AST.Expression denotes whether an expression has side effects.
+ */
+object Purity {
+
+  /**
+   * Pure indicates that an expression has no side effects.
+   */
+  case object Pure extends Purity
+
+  /**
+   * Impure indicates that an expression has side effects.
```

Comment:
Represents an impure expression (i.e. an expression that could potentially have side-effects).